File: Python/_00433_MinimumGeneticMutation.py
from collections import deque
from typing import List


class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        """
        A gene string can be represented by an 8-character long string, with
        choices from `'A'`, `'C'`, `'G'`, and `'T'`.

        Suppose we need to investigate a mutation from a gene string `startGene`
        to a gene string `endGene` where one mutation is defined as one single
        character changed in the gene string.

        - For example, `"AACCGGTT" --> "AACCGGTA"` is one mutation.

        There is also a gene bank `bank` that records all the valid gene
        mutations. A gene must be in `bank` to make it a valid gene string.

        Given the two gene strings `startGene` and `endGene` and the gene bank
        `bank`, return *the minimum number of mutations needed to mutate from*
        `startGene` *to* `endGene`. If there is no such a mutation, return `-1`.

        Note that the starting point is assumed to be valid, so it might not be
        included in the bank.

        Args:
            startGene (str): beginning 8-character gene string
            endGene (str): ending 8-character gene string
            bank (List[str]): list of valid gene mutation

        Returns:
            int: the minimum number of mutations needed to mutate from
            `startGene` to `endGene`
        """

        bank_set = set(bank) # Use set for O(1) lookups.

        # If the startGene is not the endGene and the endGene is not in the bank,
        # we know right away that there is no valid mutation path to the endGene.
        if startGene != endGene and endGene not in bank_set:
            return -1

        # queue of valid gene strings to check against the endGene and the 
        # number of mutataions required to get to each gene string
        queue = deque([(startGene, 0)])
        # set of already validated genes that will not need to be queued again
        checked = {startGene}

        # While there are valid gene strings to check against the endGene
        while queue:
            # Get the first gene string in the queue and the number of mutations
            # required to get to this gene string.
            gene_string, mutations = queue.popleft()

            # If this gene string is the endGene, we can return the mutations count.
            if gene_string == endGene:
                return mutations

            # For each gene in the gene string, try each of the other genes in 
            # this position and check for validity.
            for i in range(len(gene_string)):
                for gene in "ACGT":
                    if gene_string[i] != gene:
                        mutation = gene_string[:i] + gene + gene_string[i + 1 :]
                        
                        # If this is a valid mutation and this mutation hasn't
                        # been encountered yet add it to the queue to check 
                        # against the endGene and add it to the set of gene
                        # strings checked for validity. 
                        if mutation in bank_set and mutation not in checked:
                            checked.add(mutation)
                            queue.append((mutation, mutations + 1))
        
        # If no valid chain of mutations was found, return -1.
        return -1


import unittest


class TestSolution(unittest.TestCase):
    def setUp(self) -> None:
        self.sol = Solution()

    def test_example1(self):
        startGene = "AACCGGTT"
        endGene = "AACCGGTA"
        bank = ["AACCGGTA"]
        expected = 1
        actual = self.sol.minMutation(startGene, endGene, bank)
        self.assertEqual(expected, actual)

    def test_example2(self):
        startGene = "AACCGGTT"
        endGene = "AAACGGTA"
        bank = ["AACCGGTA", "AACCGCTA", "AAACGGTA"]
        expected = 2
        actual = self.sol.minMutation(startGene, endGene, bank)
        self.assertEqual(expected, actual)


if __name__ == "__main__":
    unittest.main()

File: JavaScript/src/_00433_MinimumGeneticMutation.js
/**
 * A gene string can be represented by an 8-character long string, with choices
 * from `'A'`, `'C'`, `'G'`, and `'T'`.
 *
 * Suppose we need to investigate a mutation from a gene string `startGene` to a
 * gene string `endGene` where one mutation is defined as one single character
 * changed in the gene string.
 *
 * - For example, `"AACCGGTT" --> "AACCGGTA"` is one mutation.
 *
 * There is also a gene bank `bank` that records all the valid gene mutations. A
 * gene must be in bank to make it a valid gene string.
 *
 * Given the two gene strings `startGene` and `endGene` and the gene bank `bank`,
 * return *the minimum number of mutations needed to mutate from* `startGene`
 * *to* `endGene`. If there is no such a mutation, return `-1`.
 *
 * Note that the starting point is assumed to be valid, so it might not be
 * included in the bank.
 * @param {string} startGene the beginning gene string
 * @param {string} endGene the ending gene string
 * @param {string[]} bank the list of valid mutation
 * @return {number} the minimum number of mutation needed to mutate from
 * `startGene` to `endGene`
 */
var minMutation = function (startGene, endGene, bank) {
    const validSet = new Set(bank); // Use a set for O(1) lookups.

    // We can return right away if the end gene is not valid or the start gene
    // is the same as the end gene.
    if (!validSet.has(endGene)) return -1;
    if (startGene === endGene) return 0;

    /** Queue of validated genes strings to compare to the end gene. */
    const queue = [[startGene, 0]];
    /** Set of alreayd validated and queued gene stings */
    const checked = new Set([startGene]);

    // While there are genes strings to compare to the end gene...
    while (queue.length > 0) {
        // Get the first gene string from the queue along with the number of
        // mutations required to result in this sequence.
        const [currentGene, mutations] = queue.shift();

        // Use an array for easier string mutation.
        const geneArray = currentGene.split("");

        // Check all single gene mutation for validity and add valid mutations 
        // to the queue to compare with the end gene.
        for (let i = 0; i < geneArray.length; i++) {
            // Save the current gene to reset the array.
            const saveGene = geneArray[i];
            for (gene of "ACGT") {
                if (geneArray[i] === gene) continue;

                geneArray[i] = gene;
                const mutation = geneArray.join("");

                // If the mutation is the end gene, count this mutation and
                // return the final result.
                if (mutation === endGene) return mutations + 1;

                // If the mutation is not valid or the mutation as already been
                // queued before, we do not need to add this gene to the queue.
                if (!validSet.has(mutation) || checked.has(mutation)) continue;

                checked.add(mutation);
                queue.push([mutation, mutations + 1]);
            }
            // Reset the gene array for the next mutation.
            geneArray[i] = saveGene;
        }
    }

    return -1;
};

module.exports = minMutation;
